experimental/dds/tree2/src/feature-libraries/editable-tree-2/context.ts

```diff
@@ -0,0 +1,147 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils";
+import {
+	IEditableForest,
+	moveToDetachedField,
+	ForestEvents,
+	FieldStoredSchema,
+	FieldKey,
+} from "../../core";
+import { ISubscribable } from "../../events";
+import { DefaultEditBuilder } from "../default-field-kinds";
+import { NodeKeyManager } from "../node-key";
+import { FieldGenerator } from "../contextuallyTyped";
+import { TypedSchemaCollection } from "../typed-schema";
+import { disposeSymbol } from "../../util";
+import { UntypedField } from "./editableTreeTypes";
+import { makeField } from "./lazyField";
+import { LazyEntity, prepareForEditSymbol } from "./lazyEntity";
+
+/**
+ * A common context of a "forest" of EditableTrees.
+ * It handles group operations like transforming cursors into anchors for edits.
+ * @alpha
+ */
+export interface TreeContext extends ISubscribable<ForestEvents> {
+	/**
+	 * Gets the root field of the tree.
+	 */
+	get root(): UntypedField;
+
+	/**
+	 * Schema used within this context.
+	 * All data must conform to these schema.
+	 */
+	readonly schema: TypedSchemaCollection;
+}
+
+/**
+ * Implementation of `EditableTreeContext`.
+ *
+ * @remarks An editor is required to edit the EditableTrees.
+ */
+export class Context implements TreeContext {
+	public readonly withCursors: Set<LazyEntity> = new Set();
+	public readonly withAnchors: Set<LazyEntity> = new Set();
+
+	private readonly eventUnregister: (() => void)[];
+
+	/**
+	 * @param forest - the Forest
+	 * @param editor - an editor that makes changes to the forest.
+	 * @param nodeKeys - an object which handles node key generation and conversion
+	 * @param nodeKeyFieldKey - an optional field key under which node keys are stored in this tree.
+	 * If present, clients may query the {@link LocalNodeKey} of a node directly via the {@link localNodeKeySymbol}.
+	 */
+	public constructor(
+		public readonly schema: TypedSchemaCollection,
+		public readonly forest: IEditableForest,
+		public readonly editor: DefaultEditBuilder,
+		public readonly nodeKeys: NodeKeyManager,
+		public readonly nodeKeyFieldKey?: FieldKey,
+	) {
+		this.eventUnregister = [
+			this.forest.on("beforeChange", () => {
+				this.prepareForEdit();
+			}),
+		];
+	}
+
+	/**
+	 * Call before editing.
+	 */
+	private prepareForEdit(): void {
+		for (const target of this.withCursors) {
+			target[prepareForEditSymbol]();
+		}
+		assert(this.withCursors.size === 0, "prepareForEdit should remove all cursors");
+	}
+
+	/**
+	 * Call to free resources.
+	 * It is invalid to use the context after this.
+	 */
+	public free(): void {
```

Comment:
that would be fine. Not required (to avoid name collisions) but fine. Since our JS target version and TS language version don't have proper support for the Special symbol.dispose, it doesn't really matter, and this is how it is since its copied from the editable-tree1 context type. Anyway, I have added IDisposable and used it and the symbol here for clarity and consistancy.